govcd/vdc.go

```diff
@@ -495,10 +498,66 @@ func (vdc *Vdc) ComposeRawVApp(name string, description string) error {
 	return nil
 }
 
+// CreateRawVApp creates an empty vApp
+func (vdc *Vdc) CreateRawVApp(name string, description string) (*VApp, error) {
+	vcomp := &types.ComposeVAppParams{
+		Ovf:         types.XMLNamespaceOVF,
+		Xsi:         types.XMLNamespaceXSI,
+		Xmlns:       types.XMLNamespaceVCloud,
+		Deploy:      false,
+		Name:        name,
+		PowerOn:     false,
+		Description: description,
+	}
+
+	vdcHref, err := url.ParseRequestURI(vdc.Vdc.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error getting vdc href: %s", err)
+	}
+	vdcHref.Path += "/action/composeVApp"
+
+	var vAppContents types.VApp
+
+	_, err = vdc.client.ExecuteRequest(vdcHref.String(), http.MethodPost,
+		types.MimeComposeVappParams, "error instantiating a new vApp:: %s", vcomp, &vAppContents)
+	if err != nil {
+		return nil, fmt.Errorf("error executing task request: %s", err)
+	}
+
+	if vAppContents.Tasks != nil {
```

Comment:
Running some tests, vAppContents.Tasks is always nil so no tasks are waited

Reply:
I guess this can really be removed. Response status 201 (Created) hints there is no task to be here. (202 is returned for tasks which actually sparks that we could possibly validate and find if there are any other historical cases - not in this PR)

```
2021/06/25 08:46:22 --------------------------------------------------------------------------------
2021/06/25 08:46:22 Request caller: govcd.(*Vdc).CreateRawVApp-->govcd.(*Vdc).CreateRawVApp-->govcd.executeRequestWithApiVersion-->govcd.
(*Client).executeRequest-->govcd.executeRequestCustomErr-->govcd.(*Client).NewRequestWithApiVersion-->govcd.(*Client).newRequest
2021/06/25 08:46:22 POST https://hostname/api/vdc/810ceb94-fd81-4088-ad98-3c6ef3ca8e56/action/composeVApp
2021/06/25 08:46:22 --------------------------------------------------------------------------------
2021/06/25 08:46:22 Request data: [348]
<?xml version="1.0" encoding="UTF-8"?>
  <ComposeVAppParams xmlns:ovf="http://schemas.dmtf.org/ovf/envelope/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http
://www.vmware.com/vcloud/v1.5" name="TestVCD.TestCreateRawVapp" deploy="false" powerOn="false">
      <Description>test compose raw app</Description>
  </ComposeVAppParams>
2021/06/25 08:46:22 Req header:
2021/06/25 08:46:22     X-Vmware-Vcloud-Token-Type: [Bearer]
2021/06/25 08:46:22     Authorization: [********]
2021/06/25 08:46:22     User-Agent: [go-vcloud-director]
2021/06/25 08:46:22     X-Vmware-Vcloud-Access-Token: [********]
2021/06/25 08:46:22     Accept: [application/*+xml;version=33.0]
2021/06/25 08:46:23 ################################################################################
2021/06/25 08:46:23 Response caller check%2ev1.(*suiteRunner).forkTest.func1-->govcd.(*TestVCD).TestCreateRawVapp-->govcd.(*Vdc).CreateRawVApp-->govcd.(*Vdc).CreateRawVApp-->govcd.(*Client).executeRequest-->govcd.decodeBody
2021/06/25 08:46:23 Response status 201 Created
2021/06/25 08:46:23 ################################################################################
```

Reply:
I provided a fix in the issue when it was raised, have compiled and tested with success.

I'm new to this language so was assuming what is above is better/optimised

Reply:
I have a different experience with this function.
If I add a deletion immediately after creation, the operation fails because the vApp is still busy being created.
Looking at the logs, I see a task, and waiting for its completion seems the right operation to do.
```
2021/06/25 17:21:25 --------------------------------------------------------------------------------
2021/06/25 17:21:25 Request caller: govcd.(*Vdc).CreateRawVApp-->govcd.(*Vdc).CreateRawVApp-->govcd.executeRequestWithApiVersion-->govcd.(*Client).executeRequest-->govcd.executeRequestCustomErr-->govcd.(*Client).NewRequestWithApiVersion-->govcd.(*Client).newRequest
2021/06/25 17:21:25 POST https://example.com/api/vdc/01da5777-bee6-4a91-a94a-2298e43d1144/action/composeVApp
2021/06/25 17:21:25 --------------------------------------------------------------------------------
2021/06/25 17:21:25 Req header:
2021/06/25 17:21:25 	X-Vmware-Vcloud-Access-Token: [********]
2021/06/25 17:21:25 	Accept: [application/*+xml;version=33.0]
2021/06/25 17:21:25 	X-Vmware-Vcloud-Token-Type: [Bearer]
2021/06/25 17:21:25 	Authorization: [********]
2021/06/25 17:21:25 	User-Agent: [go-vcloud-director]
2021/06/25 17:21:25 Request data: [348]
<?xml version="1.0" encoding="UTF-8"?>
  <ComposeVAppParams xmlns:ovf="http://schemas.dmtf.org/ovf/envelope/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.vmware.com/vcloud/v1.5" name="TestVCD.TestCreateRawVapp" deploy="false" powerOn="false">
      <Description>test compose raw app</Description>
  </ComposeVAppParams>
2021/06/25 17:21:26 ################################################################################
2021/06/25 17:21:26 Response caller check%2ev1.(*suiteRunner).forkTest.func1-->govcd.(*TestVCD).TestCreateRawVapp-->govcd.(*Vdc).CreateRawVApp-->govcd.(*Vdc).CreateRawVApp-->govcd.(*Client).executeRequest-->govcd.decodeBody
2021/06/25 17:21:26 Response status 201 Created
2021/06/25 17:21:26 ################################################################################
2021/06/25 17:21:26 Response header:
2021/06/25 17:21:26 	Cache-Control: [no-store, must-revalidate]
2021/06/25 17:21:26 	Vary: [Accept-Encoding, User-Agent]
2021/06/25 17:21:26 	Date: [Fri, 25 Jun 2021 15:21:26 GMT]
2021/06/25 17:21:26 	Content-Type: [application/vnd.vmware.vcloud.vapp+xml;version=33.0]
2021/06/25 17:21:26 	Location: [https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b]
2021/06/25 17:21:26 	X-Vmware-Vcloud-Request-Execution-Time: [260]
2021/06/25 17:21:26 	X-Vmware-Vcloud-Ceip-Id: [964ae88a-0caf-4b8b-bcec-fd0f42085c8f]
2021/06/25 17:21:26 	Content-Length: [5936]
2021/06/25 17:21:26 	X-Vmware-Vcloud-Request-Id: [96e0eee7-7997-4340-a48f-5ed6cf8a8e62]
2021/06/25 17:21:26 Response text: [5936]
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<VApp xmlns="http://www.vmware.com/vcloud/v1.5" xmlns:ovf="http://schemas.dmtf.org/ovf/envelope/1" xmlns:vssd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData" xmlns:common="http://schemas.dmtf.org/wbem/wscim/1/common" xmlns:rasd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData" xmlns:vmw="http://www.vmware.com/schema/ovf" xmlns:ovfenv="http://schemas.dmtf.org/ovf/environment/1" xmlns:vmext="http://www.vmware.com/vcloud/extension/v1.5" xmlns:ns9="http://www.vmware.com/vcloud/versions" ovfDescriptorUploaded="true" deployed="false" status="0" name="TestVCD.TestCreateRawVapp" id="urn:vcloud:vapp:333ec3a6-0b55-4b09-abf4-ed7ff57f593b" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b" type="application/vnd.vmware.vcloud.vApp+xml">
    <Link rel="up" href="https://example.com/api/vdc/01da5777-bee6-4a91-a94a-2298e43d1144" type="application/vnd.vmware.vcloud.vdc+xml"/>
    <Link rel="up" href="https://example.com/api/vdc/01da5777-bee6-4a91-a94a-2298e43d1144" type="application/vnd.vmware.vcloud.vdc+json"/>
    <Link rel="down" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/owner" type="application/vnd.vmware.vcloud.owner+xml"/>
    <Link rel="down" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/owner" type="application/vnd.vmware.vcloud.owner+json"/>
    <Link rel="down" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/metadata" type="application/vnd.vmware.vcloud.metadata+xml"/>
    <Link rel="down" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/metadata" type="application/vnd.vmware.vcloud.metadata+json"/>
    <Link rel="ovf" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/ovf" type="text/xml"/>
    <Link rel="down" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/productSections/" type="application/vnd.vmware.vcloud.productSections+xml"/>
    <Link rel="down" href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/productSections/" type="application/vnd.vmware.vcloud.productSections+json"/>
    <Description>test compose raw app</Description>
    <Tasks>
        <Task cancelRequested="false" expiryTime="2021-09-23T15:21:26.025Z" operation="Composing Virtual Application TestVCD.TestCreateRawVapp(333ec3a6-0b55-4b09-abf4-ed7ff57f593b)" operationName="vdcComposeVapp" serviceNamespace="com.vmware.vcloud" startTime="2021-06-25T15:21:26.025Z" status="queued" name="task" id="urn:vcloud:task:ff929406-d590-4e3a-adae-03dea9487942" href="https://example.com/api/task/ff929406-d590-4e3a-adae-03dea9487942" type="application/vnd.vmware.vcloud.task+xml">
            <Link rel="edit" href="https://example.com/api/task/ff929406-d590-4e3a-adae-03dea9487942" name="task" type="application/vnd.vmware.vcloud.task+xml"/>
            <Link rel="edit" href="https://example.com/api/task/ff929406-d590-4e3a-adae-03dea9487942" name="task" type="application/vnd.vmware.vcloud.task+json"/>
            <Owner href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b" id="urn:vcloud:vapp:333ec3a6-0b55-4b09-abf4-ed7ff57f593b" name="TestVCD.TestCreateRawVapp" type="application/vnd.vmware.vcloud.vApp+xml"/>
            <User href="https://example.com/api/admin/user/25a831b8-44fa-4e37-b19a-576e11e1dcb8" id="urn:vcloud:user:25a831b8-44fa-4e37-b19a-576e11e1dcb8" name="system" type="application/vnd.vmware.admin.user+xml"/>
            <Organization href="https://example.com/api/org/4b601684-788a-4770-a9af-b203c38cbcb6" id="urn:vcloud:org:4b601684-788a-4770-a9af-b203c38cbcb6" name="datacloud" type="application/vnd.vmware.vcloud.org+xml"/>
            <Progress>1</Progress>
            <Details></Details>
            <VcTaskList/>
        </Task>
    </Tasks>
    <LeaseSettingsSection href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/leaseSettingsSection/" type="application/vnd.vmware.vcloud.leaseSettingsSection+xml" ovf:required="false">
        <ovf:Info>Lease settings section</ovf:Info>
        <DeploymentLeaseInSeconds>0</DeploymentLeaseInSeconds>
        <StorageLeaseInSeconds>0</StorageLeaseInSeconds>
    </LeaseSettingsSection>
    <ovf:NetworkSection xmlns:ns10="http://www.vmware.com/vcloud/v1.5" ns10:type="application/vnd.vmware.vcloud.networkSection+xml" ns10:href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/networkSection/">
        <ovf:Info>The list of logical networks</ovf:Info>
    </ovf:NetworkSection>
    <NetworkConfigSection href="https://example.com/api/vApp/vapp-333ec3a6-0b55-4b09-abf4-ed7ff57f593b/networkConfigSection/" type="application/vnd.vmware.vcloud.networkConfigSection+xml" ovf:required="false">
        <ovf:Info>The configuration parameters for logical networks</ovf:Info>
    </NetworkConfigSection>
    <DateCreated>2021-06-25T15:21:26.014Z</DateCreated>
    <Owner type="application/vnd.vmware.vcloud.owner+xml">
        <User href="https://example.com/api/admin/user/31385982-dae3-4f29-a02c-c0f241354bf3" name="system" type="application/vnd.vmware.admin.user+xml"/>
    </Owner>
    <autoNature>false</autoNature>
    <InMaintenanceMode>false</InMaintenanceMode>
</VApp>
```

I will be happy to change if I get more info about this operation.

Reply:
I surrender :) Lets have it then